ibis/expr/types/relations.py

```diff
@@ -1993,31 +2000,34 @@ def select(
         import ibis.expr.analysis as an
         from ibis.selectors import Selector
 
-        exprs = [
-            e
-            for expr in exprs
-            for e in (
-                expr.expand(self)
-                if isinstance(expr, Selector)
-                else map(self._ensure_expr, util.promote_list(expr))
-            )
-        ]
-        exprs.extend(
+        new_exprs = []
+
+        for expr in exprs:
+            if isinstance(expr, Selector):
+                new_exprs.extend(expr.expand(self))
+            elif isinstance(expr, Mapping):
+                new_exprs.extend(
+                    _ensure_expr(self, value).name(name) for name, value in expr.items()
```

Comment:
Note that `self._ensure_expr` and `_ensure_expr` differ in how they treat str/int indices. Previously we were always using `self._ensure_expr` in this method, now this uses a mix. We should probably standardize on one here (or add a comment if the discrepancy is intentional).

```suggestion
                    self._ensure_expr(value).name(name) for name, value in expr.items()
```

Reply:
This won't work. `_ensure_expr(self, value)` handles the case of a literal, while `self._ensure_expr` does not.

This code is basically gone in `the-epic-split` FWIW.

Reply:
We might be able to move entirely over to the function form, lemme check.

Reply:
Using `_ensure_check(self, expr)` everywhere breaks a bunch of tests.

Reply:
> This won't work. `_ensure_expr(self, value)` handles the case of a literal, while `self._ensure_expr` does not.

Right. I thought we didn't want to support auto-broadcasting literals in `t.select`, only in `t.mutate`, hence the difference in the implementations here. `t.select("fizz")` should error if there's no column `fizz` in `t`, not broadcast a new column with `"fizz"` in every row (where `mutate` would do that). Following that, I don't think we want to support literal values in a mapping passed to `t.select` either. IMO how we treat the values in the mapping should be the same as how we treat the values in `**kwargs` or `*args`.

Reply:
Ah right.

What about the floating point case? Do we care? for `str`/`int` they are ambiguous with column references, but for other `float` that's not the case.

I guess if we make lack of literal support there consistent we should perhaps not allow floating point values there either?

Reply:
There's other values we could (not) allow too, like `Decimal`s.

Reply:
The main problem now is making a reasonable error message without checking for every possible thing under the sun. I suppose I could catch the attribute error when `.name` is called, but that seems less than ideal.

Reply:
Ok, I reverted back to using `self._ensure_expr`-only in `select`.

Perhaps we can live with the `AttributeError` that's thrown when passing a Python literal value like `1.0` for now?

Reply:
> I guess if we make lack of literal support there consistent we should perhaps not allow floating point values there either?

We already don't, `t.select` currently won't coerce any literals to expressions. If you pass it as an arg (`t.select(1.5)`) you get a nice error, if you do it as a kwarg (`t.select(foo=1.5)`) you get an `AttributeError` for `'name'`. I guess for now just use `self._ensure_expr` and it doesn't make anything worse, and is at least consistent? As you said things are different already in TES branch.

Reply:
SGTM, resolving.